inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/send/SenderManager.java

```diff
@@ -20,190 +20,201 @@
-            requestIdSeq.set(requestId);
-        }
-        return requestId;
+    private void reconnect(InetSocketAddress inetSocketAddress, int timeout)
+            throws IOException {
+        socket = new Socket();
+        socket.connect(inetSocketAddress, timeout);
+        socket.setSoTimeout(timeout);
     }
 
     /**
      * Send data with command
      */
-    public void send(AuditApi.BaseCommand baseCommand, AuditApi.AuditRequest auditRequest) {
+    public boolean send(AuditApi.BaseCommand baseCommand, AuditApi.AuditRequest auditRequest) {
         AuditData data = new AuditData(baseCommand, auditRequest);
-        // cache first
-        Long requestId = baseCommand.getAuditRequest().getRequestId();
-        this.dataMap.putIfAbsent(requestId, data);
-        this.sendData(data.getDataByte());
+        for (int retry = 0; retry < auditConfig.getRetryTimes(); retry++) {
+            if (sendData(data.getDataByte())) {
+                return true;
+            }
+            LOGGER.warn("Failed to send data on attempt {}. Retrying...", retry + 1);
+            sleep();
+        }
+
+        LOGGER.error("Failed to send data after {} attempts. Storing data for later retry.",
+                auditConfig.getRetryTimes());
+        failedDataMap.putIfAbsent(baseCommand.getAuditRequest().getRequestId(), data);
+        return false;
+    }
+
+    private void readFully(InputStream is, byte[] buffer, int len) throws IOException {
+        int bytesRead;
+        int totalBytesRead = 0;
+        while (totalBytesRead < len
+                && (bytesRead = is.read(buffer, totalBytesRead, len - totalBytesRead)) != -1) {
+            totalBytesRead += bytesRead;
+        }
     }
 
     /**
      * Send data byte array
      */
-    private void sendData(byte[] data) {
-        if (data == null || data.length <= 0) {
-            LOG.warn("send data is empty!");
-            return;
+    private boolean sendData(byte[] data) {
+        if (data == null || data.length == 0) {
+            LOGGER.warn("Send data is empty!");
+            return false;
         }
-        ByteBuf dataBuf = ByteBufAllocator.DEFAULT.buffer(data.length);
-        dataBuf.writeBytes(data);
-        SenderResult result = this.sender.send(dataBuf);
-        if (!result.result) {
-            this.sender.setHasSendError(true);
+        if (!checkSocket()) {
+            return false;
+        }
+        try {
+            OutputStream outputStream = socket.getOutputStream();
+            InputStream inputStream = socket.getInputStream();
+
+            outputStream.write(data);
+
+            byte[] header = new byte[PACKAGE_HEADER_LEN];
+            readFully(inputStream, header, PACKAGE_HEADER_LEN);
+
+            int bodyLen = ((header[0] & 0xFF) << 24) |
+                    ((header[1] & 0xFF) << 16) |
+                    ((header[2] & 0xFF) << 8) |
+                    (header[3] & 0xFF);
+
+            byte[] body = new byte[bodyLen];
```

Comment:
Do we need to determine if bodyLen is within a reasonable range here

Reply:
Fixed